core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java

```diff
@@ -273,27 +281,38 @@ public long getDefaultBlockSize() {
     AlluxioURI path = new AlluxioURI(HadoopUtils.getPathWithoutScheme(file.getPath()));
     List<FileBlockInfo> blocks = getFileBlocks(path);
     List<BlockLocation> blockLocations = new ArrayList<>();
+    Map<String, WorkerNetAddress> workerHosts = null;
     for (FileBlockInfo fileBlockInfo : blocks) {
       long offset = fileBlockInfo.getOffset();
       long end = offset + fileBlockInfo.getBlockInfo().getLength();
       // Check if there is any overlapping between [start, start+len] and [offset, end]
       if (end >= start && offset <= start + len) {
-        ArrayList<String> names = new ArrayList<>();
-        ArrayList<String> hosts = new ArrayList<>();
         // add the existing in-Alluxio block locations
-        for (alluxio.wire.BlockLocation location : fileBlockInfo.getBlockInfo().getLocations()) {
-          HostAndPort address = HostAndPort.fromParts(location.getWorkerAddress().getHost(),
-              location.getWorkerAddress().getDataPort());
-          names.add(address.toString());
-          hosts.add(address.getHostText());
+        Collection<WorkerNetAddress> locations = fileBlockInfo.getBlockInfo().getLocations()
+            .stream().map(alluxio.wire.BlockLocation::getWorkerAddress).collect(toList());
+        if (locations.isEmpty()) {
+          // No in-Alluxio location, fallback to use under file system locations with
+          // co-located workers.
+          if (workerHosts == null) {
+            // lazy initialization for rpc call
+            workerHosts = getHostToWorkerMap();
+          }
+          Map<String, WorkerNetAddress> finalWorkerHosts = workerHosts;
+          locations = fileBlockInfo.getUfsLocations().stream()
+              .map(location -> finalWorkerHosts.get(HostAndPort.fromString(location).getHostText()))
+              .filter(Objects::nonNull).collect(toList());
         }
-        // add under file system locations
-        for (String location : fileBlockInfo.getUfsLocations()) {
-          names.add(location);
-          hosts.add(HostAndPort.fromString(location).getHostText());
+        if (locations.isEmpty()) {
+          // Fallback to add all workers to the location so some apps (Impala) won't panic.
+          locations = workerHosts.values();
```

Comment:
This full list is returned as block locations when no cached or co-located worker is found. I don't think order matters in the return value of the API call. It might help workload distribution if the user of the API simply pick first block location to use and we randomize the list. Is that what you concerned about?

Reply:
Yes, this is my main concern.

Reply:
I see. I will shuffle the list before returning it.